### PR TITLE
WebAuthn Walk-Through

### PR DESCRIPTION
**Pull request by FFincannon**

### CODE REVIEW

**Comment by FFincannon** [comment]
@elukewalker , please read through one last time, then merge this PR. (https://yubico.atlassian.net/browse/SADEV-436)

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 1]
```diff
@@ -0,0 +1,283 @@
+== WebAuthn Walk-Through
+Build an app that enables your users to register and authenticate through WebAuthn. To implement the WebAuthn security protocol in your app, use the link:https://www.w3.org/TR/webauthn/[W3C Web Authentication API] specification and Yubico’s link:https://github.com/Yubico/java-webauthn-server/tree/master/webauthn-server-demo[Java WebAuthn Server].
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:2]: consider leaving off 'security protocol' - > To implement WebAuthn in your app

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+=== Overview
+With WebAuthn (also known as FIDO2), public-key cryptography is used to authenticate users to an online service, also known as a Relying Party (RP). When the end-user registers for an online service, an RP-specific credential key pair, i.e., a private key and a public key, is generated on the authenticator and the public key is sent to the RP (the private key never leaves the authenticator). When the user makes the request to log in, the authenticator sends an assertion that proves the user possesses the private key. The RP uses the public key to validate the assertion before allowing the user to log in.
+
+With WebAuthn also comes the concept of a resident key credential stored on the authenticator instead of encrypted and stored on the server. Resident keys enable passwordless authentication scenarios.
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:7]: rename 'resident' to discoverable

Comment (elukewalker) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc:7]: change 'passwordless' to 'usernameless'

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+Yubico’s Java WebAuthn Server demo:: link:https://github.com/Yubico/java-webauthn-server/tree/master/webauthn-server-demo[webauthn-server-demo]
+Git:: link:https://git-scm.com/[https://git-scm.com/]
+Dependency configuration:: link:https://github.com/Yubico/java-webauthn-server#dependency-configuration[dependency-configuration]: Maven or Gradle
+Javadoc API documentation:: link:https://github.com/Yubico/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/package-summary.html[package-summary] for the following:
+
+    * ``CredentialRepository`` interface,
+    * ``RelyingParty`` class and
+    * ``navigator.credentials.create()`` and
+    * ``navigator.credentials.get()`` methods.
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:23]: 1. The visual formatting of this block looks off. Can you double check it locally? 
2. An unordered list seems like a path forward.
3a. In the sub unordered list, the `` are showing up on the page. Should they be removed? 
3b. Also I suggest linking to the each of the class/interface descriptions. E.g. the CredentialRepository Interface should link to https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/CredentialRepository.html
3c. The navigator.credentials.create() and get() are not in the Java WebAuthn Server API documentation.  Those are in the W3C WebAuthn Spec.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+
+=== Resources
+
+* link:https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#authenticator-api[Client to Authenticator Protocol (CTAP) authenticator API]
+* link:https://www.w3.org/TR/webauthn/[Web Authentication Public Key Credentials API]
+* link:../Software_Projects/WebAuthn-FIDO2/WebAuthn-FIDO2_Server_Libraries/[WebAuthn FIDO2 Server Libraries]
+* link:../Software_Projects/WebAuthn-FIDO2/WebAuthn-FIDO2_Host_Libraries/[WebAuthn FIDO2 Host Libraries]
+* link:https://www.yubico.com/products/services-software/download/yubikey-manager/[YubiKey Manager]
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:32]: I suggest moving the resources to the end of the document. It seams to break the reading flow a bit.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+
+=== Run Demo
+To start the demo server, run the following in the ``java-webauthn-server`` directory:
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:36]: The `` is showing up on the the page.  Did you mean to use a single `?

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+When the server is running, use a browser supporting WebAuthn and open https://localhost:5000 to access the website.
+
+[NOTE]
+======
+You will get warnings in your browser about the connection not being secure. This is expected, because this server uses a self-signed certificate. You can safely proceed to the site.
+======
+
+
+=== Registration
+The sequence of events in the registration flow is described in detail in the “Registration flow” section of the link:WebAuthn_Developer_Guide/WebAuthn_Client_Registration.adoc[WebAuthn Client Registration chapter of the WebAuthn Developer Guide].
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:49]: I would move this text to the end of the 'Configure Client Registration Ceremony' section as wordsmith it as a resource to learn more about the registration flow.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+
+=== Authentication
+The sequence of events in the authentication flow is described in detail in the “Authentication Flow” section of the link:/WebAuthn_Developer_Guide/WebAuthn_Client_Authentication.adoc[Client Authentication chapter of the WebAuthn Developer Guide].
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:53]: I would move this text to the end of the 'Configure Client Authentication Ceremony' section as wordsmith it as a resource to learn more about the registration flow.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+
+== Client-side Development
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:56]: I suggest discarding the concept of client-side and server side and development sections. Instead, integrate the server and client steps for registration, and do likewise for authentication. I propose walking the user through the registration flow step by step which involves a series of requests and responses between the client and server.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+=== Overview
+With WebAuthn (also known as FIDO2), public-key cryptography is used to authenticate users to an online service, also known as a Relying Party (RP). When the end-user registers for an online service, an RP-specific credential key pair, i.e., a private key and a public key, is generated on the authenticator and the public key is sent to the RP (the private key never leaves the authenticator). When the user makes the request to log in, the authenticator sends an assertion that proves the user possesses the private key. The RP uses the public key to validate the assertion before allowing the user to log in.
+
+With WebAuthn also comes the concept of a resident key credential stored on the authenticator instead of encrypted and stored on the server. Resident keys enable passwordless authentication scenarios.
+
+For a detailed description of the registration and authentication ceremonies, see the “How it works” section of the link:WebAuthn_Developer_Guide/Overview.adoc[Overview] in the WebAuthn Developer Guide. In the same chapter, the sections immediately following, “FIDO2/WebAuthn application architecture” and “Key points” are also helpful.
+
+
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:11]: I would explicitly state here that we are going to walk through the Java-WebAuthn-Server's registration and authentication ceremonies and explain how the WebAuthn interaction works step-by-step.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+=== Authentication
+The sequence of events in the authentication flow is described in detail in the “Authentication Flow” section of the link:/WebAuthn_Developer_Guide/WebAuthn_Client_Authentication.adoc[Client Authentication chapter of the WebAuthn Developer Guide].
+
+
+== Client-side Development
+
+=== Configure Client Registration Ceremony
+
+*1 Client initiates request to register an authenticator* User initiates account creation, or initiates new credential registration in account settings for an existing account.
```
Comment (elukewalker) [inline content/WebAuthn/WebAuthn_Walk-Through.adoc:60]: My comment for the remainder of this section of text is to transform it to be more of a narrative. The sentences are very short and concise which makes if feel a bit choppy as I read through it. I suggest sprinkling some of the why's behind the actions being taken.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 1 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Removed some time ago

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 2 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Some time ago, 'resident' was changed to 'discoverable' when it occurs as text. We should change all the links to the Resident Key chapter in the WebAuthn guide as an operation separate to this one.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 9 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc:13]: We subsequently agreed that this text can stay as it is.

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 4 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Done some time ago

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 3 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Fixed

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 5 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Yes. Fixed

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 6 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Done

File: content/WebAuthn/WebAuthn_Walk-Through.adoc (hunk 7 above)
Comment (FFincannon) [reply content/WebAuthn/WebAuthn_Walk-Through.adoc]: Done

**Pull request merged**